### ISSUE TITLE
nix-shell changes PWD in docker privileged mode

### ISSUE THREAD

**Issue by maksar** [NixOS/nix#5816]
Running latest nixos in docker with `--privileged` flag reports $PWD as `/`
```
docker run -it --rm --privileged nixos/nix:2.5.1 sh -c "cd /etc && nix-shell -p ed --command pwd" 
```
Whereas without it – as `/etc`

Previous nixos image had different behaviour.
Both with and without `--privileged` is gives `/etc`.
```
docker run -it --rm --privileged nixos/nix:2.3.16 sh -c "cd /etc && nix-shell -p ed --command pwd"
docker run -it --rm nixos/nix:2.3.16 sh -c "cd /etc && nix-shell -p ed --command pwd"
```

This leads to strange consequences in `shellHook`-s, where user expect $PWD to be a dir from which nix-shell was launched.

Is it possible to somehow enable previous behaviour (flags, configs, etc.) on demand or by default?

**Comment by maksar** [comment]
This is probably due to sandboxing, namespaces and cgroups, but I wasn't able to find any mentions of explicit PWD setting in code (except of one non-relevant place)

**Comment by maksar** [comment]
`docker run -it --rm --privileged nixos/nix:2.5.1 sh -c "nix-channel --update && nix-env -iA nixpkgs.nix_2_5 && cd /etc && nix-shell -p ed --run pwd"` => `/`
`docker run -it --rm --privileged nixos/nix:2.5.1 sh -c "nix-channel --update && nix-env -iA nixpkgs.nix_2_4 && cd /etc && nix-shell -p ed --run pwd"` => `/etc`
`docker run -it --rm --privileged nixos/nix:2.5.1 sh -c "nix-channel --update && nix-env -iA nixpkgs.nix_2_3 && cd /etc && nix-shell -p ed --run pwd"` => `/etc`

**Comment by doronbehar** [comment]
I too experience this issue, all of a sudden my scheduled GitLab CI scripts started to fail because of this.

**Comment by starcraft66** [comment]
Experiencing the same issue using Gitlab CI runners. The nix-shell enters at `/` instead of CWD and it breaks everything from my shellhook to the pipeline step itself.

**Comment by grahamc** [comment]
I see this as well, it appears to be the same as #5853.

**Comment by kenranunderscore** [comment]
I'm trying to transition from the older `nixos/nix:2.3.12` to the `nixpkgs/nix:latest` image as base for most of our CI jobs (which in turn open a `nix-shell`) and thought I was going crazy. After some trial and error I found out that `pwd` getting "lost" is also the issue here, as described above.

(In case it might help someone: GitLab injects the directory where the files are cloned into its variable `$CI_PROJECT_DIR` (something like `/builds/<user>/<repo-name>`). In our case our build scripts are invocations similar to

```shell
nix-shell -p gnumake --run 'make -s foo'
```

Changing those to the following makes it work for now:

```shell
nix-shell -p gnumake --run "cd $CI_PROJECT_DIR && make -s foo"
```
With a `shell.nix` it should be possible to use this simple trick as well by using `cd` in a `shellHook`.)

**Comment by yorickvP** [comment]
Fixed in nix 2.8 via https://github.com/NixOS/nix/pull/6348

**Comment by maksar** [comment]
Closing the issue then.
